File: amlip_cpp/src/cpp/node/collaborative_learning/ModelManagerSenderNode.cpp
// Copyright 2023 Proyectos y Sistemas de Mantenimiento SL (eProsima).
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * @file ModelManagerSenderNode.cpp
 */

#include <cpp_utils/Log.hpp>
#include <cpp_utils/exception/InconsistencyException.hpp>

#include <cpp_utils/utils.hpp>
#include <cpp_utils/types/cast.hpp>
#include <amlip_cpp/node/collaborative_learning/ModelManagerSenderNode.hpp>

#include <dds/rpc/RPCClient.hpp>
#include <dds/Participant.hpp>
#include <dds/network_utils/topic.hpp>
#include <dds/network_utils/model_manager.hpp>

namespace eprosima {
namespace amlip {
namespace node {


ModelManagerSenderNode::ModelManagerSenderNode(
        types::AmlipIdDataType id,
        uint32_t domain_id)
    : ParentNode(id, types::NodeKind::model_sender, types::StateKind::stopped, domain_id, dds::utils::ignore_locals_domain_participant_qos(
                id.name().c_str()))
    , statistics_writer_(participant_->create_writer<types::ModelStatisticsDataType>(
                dds::utils::MODEL_STATISTICS_TOPIC_NAME,
                default_statistics_datawriter_qos()))
    , model_sender_(
        participant_->create_rpc_server<types::ModelRequestDataType,
        types::ModelReplyDataType>(dds::utils::MODEL_TOPIC_NAME))
    , running_(false)
{
    logDebug(AMLIPCPP_DDS_MODELMANAGERSENDER, "Created new ModelManager Node: " << *this << ".");
}

ModelManagerSenderNode::ModelManagerSenderNode(
        types::AmlipIdDataType id)
    : ModelManagerSenderNode(id, dds::Participant::default_domain_id())
{
    // Do nothing
}

ModelManagerSenderNode::~ModelManagerSenderNode()
{
    logDebug(AMLIPCPP_DDS_MODELMANAGERSENDER, "Destroying ModelManagerSender Node: " << *this << ".");

    // If the thread is running, stop it and join thread
    if (running_)
    {
        stop();
    }

    logDebug(AMLIPCPP_DDS_MODELMANAGERSENDER, "ModelManagerSender Node Destroyed.");
}

void ModelManagerSenderNode::publish_statistics(
        const std::string& name,
        void* data,
        const uint32_t size)
{
    //! Statistical data from models.
    types::ModelStatisticsDataType statistics(name, data, size);

    statistics.server_id(participant_->id());

    // Send statistics
    statistics_writer_->publish(statistics);
}

void ModelManagerSenderNode::publish_statistics(
        const std::string& name,
        const std::vector<types::ByteType>& data)
{
    publish_statistics(
        name,
        utils::copy_to_void_ptr(utils::cast_to_void_ptr(data.data()), data.size()),
        data.size());
}

void ModelManagerSenderNode::publish_statistics(
        const std::string& name,
        const std::string& data)
{
    publish_statistics(
        name,
        utils::copy_to_void_ptr(utils::cast_to_void_ptr(data.c_str()), data.length()),
        data.length());
}

void ModelManagerSenderNode::start(
        std::shared_ptr<ModelReplier> model_replier)
{
    logDebug(AMLIPCPP_DDS_MODELMANAGERSENDER,
            "Start processing ModelManager messages by replier in : " << *this << ".");

    if (!running_.exchange(true))
    {
        change_status_(types::StateKind::running);

        sending_thread_ = std::thread(
            &ModelManagerSenderNode::process_routine_,
            this,
            model_replier);
    }
    else
    {
        throw utils::InconsistencyException(
                  STR_ENTRY << "ModelManagerSender node " << this << " is already processing data.");
    }
}

void ModelManagerSenderNode::stop()
{
    if (running_.exchange(false))
    {
        model_sender_->stop();
        sending_thread_.join();

        change_status_(types::StateKind::stopped);
    }
}

void ModelManagerSenderNode::process_routine_(
        std::shared_ptr<ModelReplier> model_replier)
{
    while (running_)
    {
        // Wait request
        types::RpcRequestDataType<types::ModelRequestDataType> request =
                model_sender_->get_request();

        if (!running_)
        {
            break;
        }

        // Call callback
        eprosima::amlip::types::ModelReplyDataType solution =
                model_replier->fetch_model(request.data());

        types::RpcReplyDataType<types::ModelReplyDataType> reply(
            request.client_id(),
            request.task_id(),
            participant_->id(),
            std::move(solution));

        // Send reply
        model_sender_->send_reply(reply);

    }
    logDebug(AMLIPCPP_DDS_MODELMANAGERSENDER, "Finishing ModelManagerSender routine.");
}

eprosima::fastdds::dds::DataWriterQos ModelManagerSenderNode::default_statistics_datawriter_qos()
{
    eprosima::fastdds::dds::DataWriterQos qos;

    qos.durability().kind = eprosima::fastdds::dds::DurabilityQosPolicyKind::TRANSIENT_LOCAL_DURABILITY_QOS;
    qos.reliability().kind = eprosima::fastdds::dds::ReliabilityQosPolicyKind::RELIABLE_RELIABILITY_QOS;
    qos.history().kind = eprosima::fastdds::dds::HistoryQosPolicyKind::KEEP_LAST_HISTORY_QOS;
    qos.history().depth = 1;

    return qos;
}

} /* namespace node */
} /* namespace amlip */
} /* namespace eprosima */

File: amlip_py/amlip_py/node/ModelManagerSenderNode.py
# Copyright 2023 Proyectos y Sistemas de Mantenimiento SL (eProsima).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""AML-IP Model Manager Sender Node API specification."""


from amlip_py.types.AmlipIdDataType import AmlipIdDataType
from amlip_py.types.ModelReplyDataType import ModelReplyDataType
from amlip_py.types.ModelRequestDataType import ModelRequestDataType

from amlip_swig import ModelManagerSenderNode as cpp_ModelManagerSenderNode
from amlip_swig import ModelReplier as cpp_ModelReplier


class ModelReplier(cpp_ModelReplier):
    """
    Model Replier class.

    This object must be called by the ModelManagerSenderNode to execute
    the  fetch_model method with each ModelRequestDataType request that is received
    from node and must return the ModelReplyDataType reply.
    """

    def fetch_model(
            self,
            model: ModelRequestDataType) -> ModelReplyDataType:
        """
        Raise exception.

        Abstract method.
        This method should be reimplemented by child class.
        """
        raise NotImplementedError('ModelReplier.fetch_model must be specialized before use.')


class ModelReplierLambda(cpp_ModelReplier):
    """
    Custom ModelReplier supporting to create it with a lambda function.

    This object is created with a lambda function that is stored inside and used for every
    ModelRequestDataType message received.
    """

    def __init__(
            self,
            callback):
        """Construct new object with lambda as callback."""
        self.callback_ = callback
        super().__init__()

    def fetch_model(
            self,
            model: ModelRequestDataType) -> ModelReplyDataType:
        """Call internal lambda."""
        return self.callback_(model)


class ModelManagerSenderNode(cpp_ModelManagerSenderNode):
    """
    AML-IP Model Manager Sender Node.
    """

    def __init__(
            self,
            id: AmlipIdDataType,
            domain: int = None):

        """
        Create a new Model Manager Sender Node with a given id.
        Parameters
        ----------
        id : AmlipIdDataTypeD
        domain : int
        """

        # Parent class constructor
        if domain is None:
            super().__init__(id)
        else:
            super().__init__(id, domain)

    def publish_statistics(
            self,
            name: str | bytes,
            data: str | bytes) -> None:

        cpp_ModelManagerSenderNode.publish_statistics(self, name, data)

    def start(
            self,
            callback=None,
            listener: ModelReplier = None):

        # Set listener by one given or creating one for callback
        self.listener_ = None
        if listener and callback:
            raise ValueError(
                'ModelManagerSenderNode constructor expects a listener object or a callback, '
                'both given.')

        if listener:
            self.listener_ = listener

        elif callback:
            self.listener_ = ModelReplierLambda(callback)

        else:
            raise ValueError(
                'ModelManagerSenderNode constructor expects a listener object or a callback, '
                'none given.')

        cpp_ModelManagerSenderNode.start(self, self.listener_)

    def stop(
            self) -> None:

        """Stop this entity if it is running. Do nothing otherwise."""
        cpp_ModelManagerSenderNode.stop(self)

    def get_id(
            self) -> AmlipIdDataType:
        """Get AMLIP id of the node."""
        return cpp_ModelManagerSenderNode.id(self)
